### PR TITLE
Adding Support for Variadic Arguments. Deprecating support for PHP 5.5

### PR DESCRIPTION
**Pull request by ingluisjimenez**
Following on https://github.com/thephpleague/pipeline/issues/18, this PR adds support for Variadic functions to be used as with the pipeline.

Unfortunately, this would require to drop support for PHP 5.5, as this feature is only available on 5.6+.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,7 +1,6 @@
 language: php
 
 php:
-  - 5.5
```
Comment (ingluisjimenez) [inline .travis.yml:4]: Unfortunately, need to remove support for 5.5, due to the use of the variadic functions

**Comment by ingluisjimenez** [comment]
Hi @frankdejonge, thoughts on this PR?

**Comment by GreatOwl** [comment]
@ingluisjimenez interesting concept for a or on this. I'm curious what would be the specific benefit of allowing the arbitrary number of parameters over embedding those parameters in the payload.

I potentially see the risk of passing a stage you don't realize requires additional parameters and suddenly your pipeline doesn't work anymore. 

I may be short sighted on the possibilities though.

**Comment by ingluisjimenez** [comment]
The major benefit would be the ability to type hint those additional parameters, and create more robust code without the requirement of writing additional validation. Think about passing objects or any other type that can be enforced with type hinting.

As you mentioned, you could always embed those parameters in the payload, but that would require your payload to be an array, which again, removes the ability of type hinting, and would require additional validation to ensure the indexes are present and are the correct type.

Regarding a stage that doesn't support additional parameters, the good thing about PHP is that you can call a method with more arguments than what it declares, although, I remember reading an RFC to remove this "feature", but, until then, you can also add stages that only support the payload as the only parameter, and this should work as well, even though some stages could require the full list of parameters.

**Comment by GreatOwl** [comment]
I see where you are going, but this does impose a set of reqs on stages that potentially makes the stack unstable. Additionally, your use case passes the Params through pipe and not to the stage directly. Out of curiousity, does a solution like the one below begin to help in what you intended to achieve? Sorry for formatting, typed on my iPhone. 
'''php
Function getStage(...$params) {
$stage = function($payload) use ($params) {
Stuff;
}
}

**Comment by ingluisjimenez** [comment]
Not sure what would be the additional requirements that this imposes. Other than breaking support for PHP 5.5, this change is fully backwards compatible with existing pipelines/stages.

One idea that could work with the existing implementation could be passing an object (i.e. Command, Request ) which encapsulates the payload as well as any additional parameters that a given Stage could require. The object would be in charge of validating/enforcing those additional parameters, if needed. The only problem here is the object would need to be mutable, as each stage in the pipeline would have to write back the result to the Command/Request object to pass it to the next stage.

**Comment by GreatOwl** [comment]
Haha, I don't think that would be good. 

So what would happen if different stages had different parameters from other stages that also have required parameters, beside the payload?

Sorry I'm challenging this so much, I'm just having trouble seeing all the potential. Thanks for humoring my queries.

**Comment by ingluisjimenez** [comment]
> So what would happen if different stages had different parameters from other stages that also have required parameters, beside the payload?

As you control the stages in your code and you would need to make them compatible and define a signature that can accommodate for that. The stage knows about the payload structure, as it expects the workload to be in a certain format, so the stage knows for what type of interface to describe, in case it uses additional parameters.

The simplest approach, without the changes from this PR, would be to use a DTO/Command object, as I mentioned above, to encapsulate additional input arguments for the stages, if any.

**Comment by ingluisjimenez** [comment]
ping @frankdejonge

**Comment by ingluisjimenez** [comment]
maybe another ping @frankdejonge ?

**Comment by bmax** [comment]
+1

**Comment by frankdejonge** [comment]
@ingluisjimenez what practical situation requires this for you? I see that question being asked before, but the reasons are not quite clear to me. You say that if you want multiple arguments now you'd require an array and loose type-safety, but you could be passing an object which unites the two values defining its relationship and making if possible to ensure type-safety. As @GreatOwl noted before, the normal compositional model of function (as done in functional programming) would be to include the parameters in the scope of the stage/function. The test examples also don't reflect this because it shows that the dependencies for both are passed along with the payload but only one is used per function, this makes the stages dependency broader that it should be. The `__invoke` function variadic case is also not the right one because it could just have the dependencies injected.

Also, the fast that one one param can be safely be forwarded to the next stage makes it a less desirable feature. If PHP would support multiple return values, collect them and pass them on I'd merge this without hesitation.

TL;DR: I'm struggling to see the need for this apart from a "it could do this" point of view. I hope I don't come across as a nitpicker now, I always really appreciate the effort put into PR's, but projects share as much knowledge as it does code and I fear this feature makes it possible to solve valid problems the wrong way.

**Comment by ingluisjimenez** [comment]
Thanks for addressing this concern.

The case for having a function bounded to a set of parameters would work as long as the parameters are available during composition, but, if you have runtime dependencies that would only be available at the time of processing the payload, then you wouldn't be able to bind these dependencies to a given step within the pipeline. Think about a case when you need to process a form, and that form belongs to a User model, which you want to pass through the pipeline along with the payload, because the given pipeline has a few stages which make use of the said User.

My intent with this change was to leverage the existing framework to allow this without requiring the client to come up with a solution. The tests definitely could be more specific to the specific problem that I'm trying to outline, but, as mentioned in a previous comment, I also agree that a DTO ( Command or something similar ) could be used to encapsulate multiple arguments or context along with a payload.

**Comment by GreatOwl** [comment]
So an idea that could go a little ways to try to support what your trying to achieve would not quite be a command object, but could help encapsulate the additional custom logic you are trying to produce. 

If you were to utilize the already existing processor interface, you could design your own processor which wraps your stage in a custom function which maintains parameters in the payload for you, and then is in charge of dishing them out properly to the internally wrapped variadic stage. You could utilize your command object inside the wrapped function to achieve your goals here without having to break the existing contract. 

I have done similar things myself utilizing the processor interface... such as pre processing between stages any validation that I may want to use to short circuit or modify behavior of future stages. Admittedly, have not used it yet to allow variadic parameter delegation.

**Comment by ingluisjimenez** [comment]
@GreatOwl yeah, that definitely sounds like an elegant solution for the existing implementation, although, it would require coding a specific processor for each pipeline that wants to handle custom parameters.

Anyways, this was just an attempt to enhance the interface of the pipeline without breaking existing implementations, but I get the point of keeping it simple and extend functionality through other means like the processor interface.

With that said, will move on and most likely implement things using an immutable DTO that can encapsulate the multiple runtime arguments that the different stages might require in order to process a given payload.

**Comment by frankdejonge** [comment]
@ingluisjimenez thank you for being such a good sport about this 👍 and thank you for considering the alternative approaches.
